compiler/src/iree/compiler/Codegen/Common/GPU/GPUMaterializeEncoding.cpp

```diff
@@ -201,6 +230,77 @@ struct GPUSetEncodingOpLoweringConversion
+    : OpInterfaceConversionPattern<linalg::ContractionOpInterface> {
+  using OpInterfaceConversionPattern<
+      linalg::ContractionOpInterface>::OpInterfaceConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(linalg::ContractionOpInterface op, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto linalgOp = cast<linalg::LinalgOp>(op.getOperation());
+    auto inputs = linalgOp.getDpsInputOperands();
+    auto outputs = linalgOp.getDpsInits();
+    auto lhsType = cast<RankedTensorType>(inputs[0]->get().getType());
+    auto rhsType = cast<RankedTensorType>(inputs[1]->get().getType());
+    auto resultType = cast<RankedTensorType>(outputs[0].getType());
+    auto lhsEncoding = IREE::Encoding::getEncodingAttr(lhsType);
+    auto rhsEncoding = IREE::Encoding::getEncodingAttr(rhsType);
+    auto resultEncoding = IREE::Encoding::getEncodingAttr(resultType);
+    if (!lhsEncoding || !rhsEncoding || !resultEncoding) {
+      LLVM_DEBUG(llvm::dbgs() << "expect encodings on operand types\n");
+      return failure();
+    }
+
+    std::optional<IREE::GPU::MmaInterfaceAttr> kind =
+        getMmaKind(lhsType.getElementType(), rhsType.getElementType(),
+                   resultType.getElementType());
+    if (!kind) {
+      LLVM_DEBUG(llvm::dbgs() << "can't find supported Mma intrinsic\n");
+      return failure();
+    }
+    LLVM_DEBUG(llvm::dbgs() << "Kind: " << kind.value() << "\n");
+
+    FailureOr<linalg::ContractionDimensions> contractionDims =
+        linalg::inferContractionDims(linalgOp);
+    assert(
+        succeeded(contractionDims) &&
+        "should always be able to infer contraction dims for contraction ops");
+   // TODO(hanchung): Support batch gemms.
+    if (!contractionDims->batch.empty()) {
+      LLVM_DEBUG(llvm::dbgs() << "batch gemm is not yet implemented\n");
+      return failure();
+    }
+
+    // TODO(hanchung): Support unrolling cases. We likely need to teach
+    // multi_mma op about interleaving K dimension.
+    MLIRContext *ctx = rewriter.getContext();
+    AffineExpr d0, d1, d2;
+    bindDims(ctx, d0, d1, d2);
+    AffineExpr mExpr = rewriter.getAffineDimExpr(0);
+    AffineExpr nExpr = rewriter.getAffineDimExpr(1);
+    AffineExpr kExpr = rewriter.getAffineDimExpr(2);
+
+    // The outer dims are all in row-major fasion after relayout.
+    auto lhsMap = AffineMap::get(3, 0, {mExpr, kExpr}, ctx);
+    auto rhsMap = AffineMap::get(3, 0, {nExpr, kExpr}, ctx);
+    auto accMap = AffineMap::get(3, 0, {mExpr, nExpr}, ctx);
+
+    SmallVector<utils::IteratorType> iteratorTypes =
+        linalgOp.getIteratorTypesArray();
+
+    // TODO(hanchung): Support batch gemms.
+    Location loc = op.getLoc();
+    SmallVector<int64_t> identityPerm = {0, 1};
+    auto mmaOp = rewriter.create<IREE::GPU::MultiMmaOp>(
+        loc, operands[0], operands[1], operands[2],
+        ArrayRef<AffineMap>{lhsMap, rhsMap, accMap},
+        iteratorTypes, kind.value());
```

Comment:
The materialization patterns expect a column-major RHS inner tile layout, right? If that is the case, then I think the MultiMmaOp needs to have an `rhs_permutation`. We also need to account for permutations introduced by data tiling for each operand. The semantics of the op are that with no permutations, each input is assumed to have a row-major inner tile layout.

You can see in this test:
https://github.com/iree-org/iree/blob/3901e6276a09f70bd3ceda0ca270df6ebf387780/compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/test/iree_gpu_ops.mlir#L190-L209
The op has `rhs_permutation = array<i64: 1, 0>` when the rhs inner tile (`32x8`) is column-major (MxK). In the case of data tiling, the inner tile will also be expanded and permuted with some general tile swizzle. It may look something like this after the expand_shape:
```mlir
#contraction_accesses = [
 affine_map<(i, j, k) -> (i, k)>,
 affine_map<(i, j, k) -> (k, j)>,
 affine_map<(i, j, k) -> (i, j)>
]
func.func @tensor_subgroup_matmul_transpose_b_32x32x8_multi_mma(
  %lhs: tensor<?x?x32x2x4xf16>,
  %rhs: tensor<?x?x32x2x4xf16>,
  %acc: tensor<?x?x4x2x4x32xf32>) -> tensor<?x?x4x2x4x32xf32> {
  %0 = iree_gpu.multi_mma %lhs, %rhs, %acc {
    indexing_maps = #contraction_accesses,
    iterator_types = [#iree_gpu.iterator_type<parallel>, #iree_gpu.iterator_type<parallel>, #iree_gpu.iterator_type<reduction>],
    kind = #iree_gpu.mma_layout<MFMA_F32_32x32x8_F16>,
    rhs_permutation = array<i64: 2, 0, 1>
  } : tensor<?x?x32x2x4xf16>, tensor<?x?x32x2x4xf16> into tensor<?x?x4x2x4x32xf32>
  return %0 : tensor<?x?x4x2x4x32xf32>
}
```
Now, `rhs_permutation = array<i64: 2, 0, 1>`, because the canonical layout after the expand_shape would be `2x4x32`, but it is `32x2x4`. If there is an additional transpose due to data tiling, then that transpose would also need to be composed with `rhs_permutation` here.

The materialization pattern should consider the transposes that happened on the lhs, rhs, and acc, and express them accordingly in the MultiMmaOp.

**Side note:** This is sort of contradictory to the assumption we had been using in our discussions that the inner tile dimensions are effectively opaque, and I believe it is part of the reason why we may need to extend the semantics (or mma enums) of the multi_mma op in order to do interleaving. Right now, the inner tile dims (of a single intrinsic tile) are separated from the outer tiles, but in data tiling we want to interleave them.

@qedawkins can you confirm that I am getting all the semantics correct?

Reply:
Interesting, and thanks for pointing it out. I think you're talking about C-Matrix, but not B-Matrix in the example? I agree that we need to attach those permutation because we need a way to express different inner layout. So yes, here we want to add `{1, 0}` to LHS, RHS, and ACC.

I can't really interpret the case of `rhs_permutation = array<i64: 2, 0, 1>` because they are defined for `MNK` mapping. However, here the rank is 3 on RHS. I feel that something is missing, and we might need to extend the semantics in this case.

https://github.com/iree-org/iree/blob/868f41e4809af70a79d4d96d41305b3c773b4079/compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUOps.td#L115-L118

After thinking a lot, I think I made a mistake in the shapes. It should be `*x4x16` for LHS after materialization but not `*x16x4`. Because we already relayout inner tiles to column-major fashion. I'll rebase on https://github.com/iree-org/iree/pull/18135 and reuse the helpers.